_posts/2025-01-08-working-with-graph-databases-in-python.md

```diff
@@ -0,0 +1,278 @@
+The URIs are what allows us to create new connections between nodes in the graph,
+and even connections between multiple graphs.
+For this to work, data is stored in `triples`. In a triple, we have 3 pieces of information:
+- Subject: the row in a typical database, referring to one feature (URI)
+- Predicate: the column in a typical database, referring to one attribute name (URI)
+- Object: the value found at the given row/column, the attribute value (URI or literal)
+
+The predicate URI should also be a subject itself, containing more triples that describe its definition.
+This means that all of the context behind the data is described within the graph database,
+an advantage which cannot be done in a relational database.
+
+Let's apply this concept to a lithology example. If we want to represent Rhyolite's lithology classification
+as an Igneous Rock in a triple, we could do the following:
+- Subject: URI to Rhyolite
+- Predicate: URI to Parent Definition
+- Object: URI to Igneous Rock
+
+The value from this data structure comes from the fact that in this example, we could find the predicates
+of Igneous Rock using its own URI, and then continue to another subject after that.
+
+The nature of `triples` also means that when you iterate over the `triples` in a
+graph database, you get a single predicate object for a single subject at a time.
+You do not get all of the predicates for a subject at once.
+Therefore, you will likely want to parse the data into a different structure
+to suit your requirements, depending on what you wish to do with the data.
+
+It is important to note that because predicates are defined using a URI,
+we can use pre-defined predicates from online `namespaces`.
+In the example below, we will see many attributes from the online `skos` namespace,
+such as `skos:prefLabel` which is essentially a human readable name,
+and `skos:broader` which is a parent node.
+
+You can find more information about the `skos` data model [here](https://www.w3.org/2009/08/skos-reference/skos.html).
+
+## Why are we working with graph databases?
+
+We started with hundreds of lithology classifications which were going to be used in a
+new Field Data Capture system, where users could input spatial lithology data.
+We wanted to plot coloured points onto a map within the system,
+so that a user could see all of their spatial lithology data visually, as this would allow them
+to quickly make sense of their data.
+However, applying hundreds of different colours to points on a map would make it very hard
+to see differences between the colours, as they would be very similar.
+Therefore, we wanted to select 50~ high level classifications,
+so that every lithology could be attributed to one of these.
+Finally, these 50~ classifications could be colour coded using colours which are more clearly different.
+
+To do this, we decided to use the data from a graph database provided by the
```

Comment:
To do this, we decided to use the data from the [simpleLithology]([http://resource.geosciml.org/classifier/cgi/lithology](https://cgi.vocabs.ga.gov.au/object?uri=http%3A//resource.geosciml.org/classifier/cgi/lithology)) vocabulary provided by the [Geoscience Terminology Working Group of the IUGS Commission for the Management and Application of Geoscience Information](https://cgi-iugs.org/project/geoscienceterminology/), a group that BGS have had a long running involvement with.